File: src/routes/ask.js
const express = require('express');
const geminiService = require('../services/gemini');
const { authenticateToken } = require('../middleware/auth');
const { db } = require('../services/firebase');

const router = express.Router();

// Health check endpoint
router.get('/health', async (req, res) => {
  try {
    const isHealthy = await geminiService.healthCheck();
    const modelInfo = geminiService.getModelInfo();
    
    res.json({
      status: isHealthy ? 'healthy' : 'unhealthy',
      timestamp: new Date().toISOString(),
      ...modelInfo
    });
  } catch (error) {
    res.status(500).json({
      status: 'error',
      error: error.message,
      timestamp: new Date().toISOString()
    });
  }
});

// Test endpoint without authentication
router.post('/test', async (req, res) => {
  try {
    const { prompt = 'Hello, this is a test message.' } = req.body;
    
    console.log('Test endpoint called with prompt:', prompt);
    
    const response = await geminiService.generateText(prompt);
    
    res.json({
      response,
      timestamp: new Date().toISOString(),
      model: 'test-mode',
      success: true
    });
  } catch (error) {
    console.error('Test endpoint error:', error);
    res.json({
      response: `Test response for: "${prompt}". The AI service is currently being optimized.`,
      timestamp: new Date().toISOString(),
      model: 'fallback',
      success: false,
      error: error.message
    });
  }
});

// Main ask endpoint with authentication
router.post('/', authenticateToken, async (req, res) => {
  try {
    const { prompt, type = 'text', imageData } = req.body;
    
    if (!prompt) {
      return res.status(400).json({ 
        error: 'Prompt is required',
        timestamp: new Date().toISOString()
      });
    }

    console.log(`Processing ${type} request for user: ${req.user.uid}`);
    console.log(`Prompt: ${prompt.substring(0, 100)}...`);

    let response;
    let processingTime = Date.now();
    
    switch (type) {
      case 'image':
        if (!imageData) {
          return res.status(400).json({ 
            error: 'Image data required for image analysis',
            timestamp: new Date().toISOString()
          });
        }
        response = await geminiService.generateFromImage(prompt, imageData);
        break;
        
      case 'code':
        response = await geminiService.generateCode(prompt);
        break;
        
      case 'search':
        response = await geminiService.generateSearch(prompt);
        break;
        
      default:
        response = await geminiService.generateText(prompt);
    }

    processingTime = Date.now() - processingTime;
    console.log(`Response generated in ${processingTime}ms`);

    // Save to history with enhanced metadata
    try {
      await db.collection('chat_history').add({
        userId: req.user.uid,
        prompt,
        response,
        type,
        timestamp: new Date(),
        processingTime,
        model: 'gemini-2.0-flash-exp',
        success: true
      });
    } catch (dbError) {
      console.error('Failed to save to history:', dbError.message);
      // Don't fail the request if history save fails
    }

    res.json({ 
      response,
      timestamp: new Date().toISOString(),
      processingTime,
      type,
      success: true
    });

  } catch (error) {
    console.error('Ask endpoint error:', error);
    
    const errorResponse = {
      error: error.message,
      timestamp: new Date().toISOString(),
      type: req.body.type || 'text',
      success: false
    };

    // Try to save error to history
    try {
      if (req.user) {
        await db.collection('chat_history').add({
          userId: req.user.uid,
          prompt: req.body.prompt,
          response: `Error: ${error.message}`,
          type: req.body.type || 'text',
          timestamp: new Date(),
          success: false,
          error: error.message
        });
      }
    } catch (dbError) {
      console.error('Failed to save error to history:', dbError.message);
    }

    // Return appropriate status code based on error type
    if (error.message.includes('API key') || error.message.includes('authentication')) {
      res.status(401).json(errorResponse);
    } else if (error.message.includes('quota') || error.message.includes('limit')) {
      res.status(429).json(errorResponse);
    } else if (error.message.includes('model') || error.message.includes('not found')) {
      res.status(503).json(errorResponse);
    } else {
      res.status(500).json(errorResponse);
    }
  }
});

// Streaming endpoint for real-time responses
router.post('/stream', authenticateToken, async (req, res) => {
  try {
    const { prompt, type = 'text' } = req.body;
    
    if (!prompt) {
      return res.status(400).json({ error: 'Prompt is required' });
    }

    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'Connection': 'keep-alive',
      'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Headers': 'Cache-Control'
    });

    // Send initial event
    res.write(`data: ${JSON.stringify({ type: 'start', timestamp: new Date().toISOString() })}\n\n`);

    try {
      const response = await geminiService.generateText(prompt);
      
      // Simulate streaming by sending chunks
      const words = response.split(' ');
      let currentText = '';
      
      for (let i = 0; i < words.length; i++) {
        currentText += (i > 0 ? ' ' : '') + words[i];
        
        res.write(`data: ${JSON.stringify({ 
          type: 'chunk', 
          content: currentText,
          progress: (i + 1) / words.length
        })}\n\n`);
        
        // Small delay to simulate real streaming
        await new Promise(resolve => setTimeout(resolve, 50));
      }
      
      // Send completion event
      res.write(`data: ${JSON.stringify({ 
        type: 'complete', 
        content: response,
        timestamp: new Date().toISOString()
      })}\n\n`);
      
    } catch (error) {
      res.write(`data: ${JSON.stringify({ 
        type: 'error', 
        error: error.message,
        timestamp: new Date().toISOString()
      })}\n\n`);
    }
    
    res.end();
    
  } catch (error) {
    console.error('Stream endpoint error:', error);
    res.status(500).json({ error: error.message });
  }
});

module.exports = router;

File: routers/ask.py
from fastapi import APIRouter, HTTPException, Depends, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
from typing import Optional
import json
import asyncio
from datetime import datetime
import logging

from middleware.auth import get_current_user, get_optional_user
from services.gemini_service import gemini_service
from services.firebase_service import get_firestore_client
from slowapi import Limiter
from slowapi.util import get_remote_address

logger = logging.getLogger(__name__)
router = APIRouter()
limiter = Limiter(key_func=get_remote_address)

class AskRequest(BaseModel):
    prompt: str
    type: str = "text"
    imageData: Optional[str] = None

class TestRequest(BaseModel):
    prompt: str = "Hello, this is a test message."

@router.get("/health")
async def health_check():
    """Health check endpoint"""
    try:
        is_healthy = await gemini_service.health_check()
        model_info = gemini_service.get_model_info()
        
        return {
            "status": "healthy" if is_healthy else "unhealthy",
            "timestamp": datetime.now().isoformat(),
            **model_info
        }
    except Exception as e:
        logger.error(f"Health check error: {e}")
        return {
            "status": "error",
            "error": str(e),
            "timestamp": datetime.now().isoformat()
        }

@router.post("/test")
async def test_endpoint(request: TestRequest):
    """Test endpoint without authentication"""
    try:
        logger.info(f"Test endpoint called with prompt: {request.prompt}")
        
        response = await gemini_service.generate_text(request.prompt)
        
        return {
            "response": response,
            "timestamp": datetime.now().isoformat(),
            "model": "test-mode",
            "success": True
        }
    except Exception as e:
        logger.error(f"Test endpoint error: {e}")
        return {
            "response": f"Test response for: \"{request.prompt}\". The AI service is currently being optimized.",
            "timestamp": datetime.now().isoformat(),
            "model": "fallback",
            "success": False,
            "error": str(e)
        }

@router.post("/")
@limiter.limit("100/15minutes")
async def ask_endpoint(request_data: Request, ask_request: AskRequest, user=Depends(get_current_user)):
    """Main ask endpoint with authentication"""
    try:
        if not ask_request.prompt:
            raise HTTPException(status_code=400, detail="Prompt is required")

        logger.info(f"Processing {ask_request.type} request for user: {user['uid']}")
        logger.info(f"Prompt: {ask_request.prompt[:100]}...")

        processing_start = datetime.now()
        
        # Generate response based on type
        if ask_request.type == "image":
            if not ask_request.imageData:
                raise HTTPException(status_code=400, detail="Image data required for image analysis")
            response = await gemini_service.generate_from_image(ask_request.prompt, ask_request.imageData)
        elif ask_request.type == "code":
            response = await gemini_service.generate_code(ask_request.prompt)
        elif ask_request.type == "search":
            response = await gemini_service.generate_search(ask_request.prompt)
        else:
            response = await gemini_service.generate_text(ask_request.prompt)

        processing_time = (datetime.now() - processing_start).total_seconds() * 1000
        logger.info(f"Response generated in {processing_time:.0f}ms")

        # Save to history
        try:
            db = get_firestore_client()
            await asyncio.to_thread(
                db.collection('chat_history').add,
                {
                    'userId': user['uid'],
                    'prompt': ask_request.prompt,
                    'response': response,
                    'type': ask_request.type,
                    'timestamp': datetime.now(),
                    'processingTime': processing_time,
                    'model': 'gemini-2.0-flash-exp',
                    'success': True
                }
            )
        except Exception as db_error:
            logger.error(f"Failed to save to history: {db_error}")

        return {
            "response": response,
            "timestamp": datetime.now().isoformat(),
            "processingTime": processing_time,
            "type": ask_request.type,
            "success": True
        }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Ask endpoint error: {e}")
        
        # Try to save error to history
        try:
            db = get_firestore_client()
            await asyncio.to_thread(
                db.collection('chat_history').add,
                {
                    'userId': user['uid'],
                    'prompt': ask_request.prompt,
                    'response': f"Error: {str(e)}",
                    'type': ask_request.type,
                    'timestamp': datetime.now(),
                    'success': False,
                    'error': str(e)
                }
            )
        except Exception as db_error:
            logger.error(f"Failed to save error to history: {db_error}")

        # Return appropriate status code based on error type
        if "API key" in str(e) or "authentication" in str(e):
            raise HTTPException(status_code=401, detail=str(e))
        elif "quota" in str(e) or "limit" in str(e):
            raise HTTPException(status_code=429, detail=str(e))
        elif "model" in str(e) or "not found" in str(e):
            raise HTTPException(status_code=503, detail=str(e))
        else:
            raise HTTPException(status_code=500, detail=str(e))

@router.post("/stream")
async def stream_endpoint(ask_request: AskRequest, user=Depends(get_current_user)):
    """Streaming endpoint for real-time responses"""
    try:
        if not ask_request.prompt:
            raise HTTPException(status_code=400, detail="Prompt is required")

        async def generate_stream():
            try:
                # Send initial event
                yield f"data: {json.dumps({'type': 'start', 'timestamp': datetime.now().isoformat()})}\n\n"
                
                response = await gemini_service.generate_text(ask_request.prompt)
                
                # Simulate streaming by sending chunks
                words = response.split(' ')
                current_text = ''
                
                for i, word in enumerate(words):
                    current_text += (' ' if i > 0 else '') + word
                    
                    yield f"data: {json.dumps({
                        'type': 'chunk',
                        'content': current_text,
                        'progress': (i + 1) / len(words)
                    })}\n\n"
                    
                    await asyncio.sleep(0.05)  # Small delay
                
                # Send completion event
                yield f"data: {json.dumps({
                    'type': 'complete',
                    'content': response,
                    'timestamp': datetime.now().isoformat()
                })}\n\n"
                
            except Exception as e:
                yield f"data: {json.dumps({
                    'type': 'error',
                    'error': str(e),
                    'timestamp': datetime.now().isoformat()
                })}\n\n"

        return StreamingResponse(
            generate_stream(),
            media_type="text/event-stream",
            headers={
                "Cache-Control": "no-cache",
                "Connection": "keep-alive",
                "Access-Control-Allow-Origin": "*",
                "Access-Control-Allow-Headers": "Cache-Control"
            }
        )
        
    except Exception as e:
        logger.error(f"Stream endpoint error: {e}")
        raise HTTPException(status_code=500, detail=str(e))